File: api/pagespeed.py
"""
Google PageSpeed Insights proxy for Atlas.
Returns a slim report for the project homepage (or another public http(s) URL).
Optional PAGESPEED_API_KEY raises Google quota; the API works without a key too.
"""
from __future__ import annotations

import json
import os
import urllib.error
import urllib.parse
import urllib.request
from typing import Any

GOOGLE_ENDPOINT = "https://www.googleapis.com/pagespeedonline/v5/runPagespeed"
ALLOWED_STRATEGIES = frozenset({"mobile", "desktop"})
CATEGORIES = ("performance", "accessibility", "best-practices", "seo")
TIMEOUT_S = 90

_FORBIDDEN_HOSTS = frozenset({"localhost", "127.0.0.1", "0.0.0.0", "::1"})


def has_api_key() -> bool:
    return bool((os.environ.get("PAGESPEED_API_KEY") or "").strip())


def _is_private_host(host: str) -> bool:
    h = (host or "").lower().strip()
    if not h or h in _FORBIDDEN_HOSTS or h.endswith(".local") or h.endswith(".internal"):
        return True
    parts = h.split(".")
    if len(parts) == 4 and all(p.isdigit() for p in parts):
        a, b = int(parts[0]), int(parts[1])
        if a == 10 or a == 127:
            return True
        if a == 192 and b == 168:
            return True
        if a == 172 and 16 <= b <= 31:
            return True
        if a == 169 and b == 254:
            return True
    return False


def validate_url(raw: str) -> str | None:
    t = (raw or "").strip()
    if not t:
        return None
    try:
        u = urllib.parse.urlparse(t)
    except Exception:
        return None
    if u.scheme not in ("http", "https"):
        return None
    host = (u.hostname or "").lower()
    if not host or "." not in host or _is_private_host(host):
        return None
    return u.geturl()


def _audit(audits: dict, key: str) -> dict | None:
    a = audits.get(key) or {}
    if not a:
        return None
    return {
        "display": a.get("displayValue") or "",
        "value": a.get("numericValue"),
        "score": a.get("score"),
    }


def _cat_score(cats: dict, key: str) -> int | None:
    s = (cats.get(key) or {}).get("score")
    if s is None:
        return None
    try:
        return int(round(float(s) * 100))
    except (TypeError, ValueError):
        return None


def _field_metric(metrics: dict, key: str) -> dict | None:
    m = metrics.get(key) or {}
    if not m:
        return None
    return {
        "percentile": m.get("percentile"),
        "category": m.get("category"),
    }


def slim_report(psi: dict[str, Any], strategy: str) -> dict[str, Any]:
    lh = psi.get("lighthouseResult") or {}
    cats = lh.get("categories") or {}
    audits = lh.get("audits") or {}
    crux = (psi.get("loadingExperience") or {}).get("metrics") or {}
    return {
        "ok": True,
        "url": psi.get("id") or "",
        "strategy": strategy,
        "fetchedAt": psi.get("analysisUTCTimestamp") or lh.get("fetchTime") or "",
        "scores": {
            "performance": _cat_score(cats, "performance"),
            "accessibility": _cat_score(cats, "accessibility"),
            "bestPractices": _cat_score(cats, "best-practices"),
            "seo": _cat_score(cats, "seo"),
        },
        "metrics": {
            "fcp": _audit(audits, "first-contentful-paint"),
            "lcp": _audit(audits, "largest-contentful-paint"),
            "tbt": _audit(audits, "total-blocking-time"),
            "cls": _audit(audits, "cumulative-layout-shift"),
            "si": _audit(audits, "speed-index"),
            "tti": _audit(audits, "interactive"),
            "inp": _audit(audits, "interaction-to-next-paint"),
        },
        "field": {
            "lcp": _field_metric(crux, "LARGEST_CONTENTFUL_PAINT_MS"),
            "inp": _field_metric(crux, "INTERACTION_TO_NEXT_PAINT")
            or _field_metric(crux, "FIRST_INPUT_DELAY_MS"),
            "cls": _field_metric(crux, "CUMULATIVE_LAYOUT_SHIFT_SCORE"),
        },
    }


def run(url: str, strategy: str = "mobile") -> dict[str, Any]:
    clean = validate_url(url)
    if not clean:
        return {"ok": False, "status": 400, "error": "A public http(s) website URL is required"}
    strat = (strategy or "mobile").strip().lower()
    if strat not in ALLOWED_STRATEGIES:
        strat = "mobile"

    params = [("url", clean), ("strategy", strat)]
    for cat in CATEGORIES:
        params.append(("category", cat))
    key = (os.environ.get("PAGESPEED_API_KEY") or "").strip()
    if key:
        params.append(("key", key))

    endpoint = GOOGLE_ENDPOINT + "?" + urllib.parse.urlencode(params)
    req = urllib.request.Request(endpoint, method="GET", headers={"Accept": "application/json"})
    try:
        with urllib.request.urlopen(req, timeout=TIMEOUT_S) as resp:
            raw = json.loads(resp.read().decode("utf-8"))
    except urllib.error.HTTPError as e:
        body = e.read().decode("utf-8", errors="replace")
        msg = f"PageSpeed API error ({e.code})"
        try:
            err = json.loads(body)
            msg = (err.get("error") or {}).get("message") or msg
        except Exception:
            pass
        return {"ok": False, "status": e.code if e.code >= 400 else 502, "error": msg}
    except Exception as e:
        return {"ok": False, "status": 502, "error": f"PageSpeed request failed ({e})"}

    if not isinstance(raw, dict) or not raw.get("lighthouseResult"):
        err = (raw.get("error") or {}).get("message") if isinstance(raw, dict) else None
        return {"ok": False, "status": 502, "error": err or "PageSpeed returned no Lighthouse result"}
    return slim_report(raw, strat)

File: netlify/functions/lib/pagespeed.js
/**
 * Google PageSpeed Insights proxy helpers (mirrors api/pagespeed.py).
 */
const GOOGLE_ENDPOINT = "https://www.googleapis.com/pagespeedonline/v5/runPagespeed";
const ALLOWED_STRATEGIES = new Set(["mobile", "desktop"]);
const CATEGORIES = ["performance", "accessibility", "best-practices", "seo"];
const TIMEOUT_MS = 90_000;
const FORBIDDEN_HOSTS = new Set(["localhost", "127.0.0.1", "0.0.0.0", "::1"]);

function hasApiKey() {
  return Boolean((process.env.PAGESPEED_API_KEY || "").trim());
}

function isPrivateHost(host) {
  const h = String(host || "").toLowerCase().trim();
  if (!h || FORBIDDEN_HOSTS.has(h) || h.endsWith(".local") || h.endsWith(".internal")) return true;
  const parts = h.split(".");
  if (parts.length === 4 && parts.every((p) => /^\d+$/.test(p))) {
    const a = Number(parts[0]);
    const b = Number(parts[1]);
    if (a === 10 || a === 127) return true;
    if (a === 192 && b === 168) return true;
    if (a === 172 && b >= 16 && b <= 31) return true;
    if (a === 169 && b === 254) return true;
  }
  return false;
}

function validateUrl(raw) {
  const t = String(raw || "").trim();
  if (!t) return null;
  try {
    const u = new URL(t);
    if (u.protocol !== "http:" && u.protocol !== "https:") return null;
    const host = (u.hostname || "").toLowerCase();
    if (!host || !host.includes(".") || isPrivateHost(host)) return null;
    return u.href;
  } catch (_) {
    return null;
  }
}

function audit(audits, key) {
  const a = (audits && audits[key]) || null;
  if (!a) return null;
  return {
    display: a.displayValue || "",
    value: a.numericValue,
    score: a.score,
  };
}

function catScore(cats, key) {
  const s = cats && cats[key] && cats[key].score;
  if (s == null || Number.isNaN(Number(s))) return null;
  return Math.round(Number(s) * 100);
}

function fieldMetric(metrics, key) {
  const m = (metrics && metrics[key]) || null;
  if (!m) return null;
  return { percentile: m.percentile, category: m.category };
}

function slimReport(psi, strategy) {
  const lh = psi.lighthouseResult || {};
  const cats = lh.categories || {};
  const audits = lh.audits || {};
  const crux = (psi.loadingExperience && psi.loadingExperience.metrics) || {};
  return {
    ok: true,
    url: psi.id || "",
    strategy,
    fetchedAt: psi.analysisUTCTimestamp || lh.fetchTime || "",
    scores: {
      performance: catScore(cats, "performance"),
      accessibility: catScore(cats, "accessibility"),
      bestPractices: catScore(cats, "best-practices"),
      seo: catScore(cats, "seo"),
    },
    metrics: {
      fcp: audit(audits, "first-contentful-paint"),
      lcp: audit(audits, "largest-contentful-paint"),
      tbt: audit(audits, "total-blocking-time"),
      cls: audit(audits, "cumulative-layout-shift"),
      si: audit(audits, "speed-index"),
      tti: audit(audits, "interactive"),
      inp: audit(audits, "interaction-to-next-paint"),
    },
    field: {
      lcp: fieldMetric(crux, "LARGEST_CONTENTFUL_PAINT_MS"),
      inp: fieldMetric(crux, "INTERACTION_TO_NEXT_PAINT")
        || fieldMetric(crux, "FIRST_INPUT_DELAY_MS"),
      cls: fieldMetric(crux, "CUMULATIVE_LAYOUT_SHIFT_SCORE"),
    },
  };
}

async function run(url, strategy) {
  const clean = validateUrl(url);
  if (!clean) {
    return { ok: false, status: 400, error: "A public http(s) website URL is required" };
  }
  let strat = String(strategy || "mobile").trim().toLowerCase();
  if (!ALLOWED_STRATEGIES.has(strat)) strat = "mobile";

  const params = new URLSearchParams();
  params.set("url", clean);
  params.set("strategy", strat);
  CATEGORIES.forEach((c) => params.append("category", c));
  const key = (process.env.PAGESPEED_API_KEY || "").trim();
  if (key) params.set("key", key);

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), TIMEOUT_MS);
  try {
    const res = await fetch(`${GOOGLE_ENDPOINT}?${params.toString()}`, {
      headers: { Accept: "application/json" },
      signal: controller.signal,
    });
    const raw = await res.json().catch(() => ({}));
    if (!res.ok) {
      const msg = (raw.error && raw.error.message) || `PageSpeed API error (${res.status})`;
      return { ok: false, status: res.status >= 400 ? res.status : 502, error: msg };
    }
    if (!raw || !raw.lighthouseResult) {
      const msg = (raw.error && raw.error.message) || "PageSpeed returned no Lighthouse result";
      return { ok: false, status: 502, error: msg };
    }
    return slimReport(raw, strat);
  } catch (err) {
    const aborted = err && (err.name === "AbortError" || /aborted/i.test(String(err)));
    return {
      ok: false,
      status: 502,
      error: aborted ? "PageSpeed timed out — try again in a moment" : `PageSpeed request failed (${err.message || err})`,
    };
  } finally {
    clearTimeout(timer);
  }
}

function jsonResponse(statusCode, payload) {
  return {
    statusCode,
    headers: {
      "Content-Type": "application/json; charset=utf-8",
      "Cache-Control": "no-store",
    },
    body: JSON.stringify(payload),
  };
}

module.exports = {
  hasApiKey,
  run,
  slimReport,
  validateUrl,
  jsonResponse,
};
